### ISSUE TITLE
Upgrade python CherryPy and other Python3 packages

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11080]
**Impact of the new feature**
WMCore in general

**Is your feature request related to a problem? Please describe.**
Given that in the beginning of 2022 we decided to drop support to python2 in WMCore, we should revisit the list of third-party python libraries and update them to the latest stable release. 

**Describe the solution you'd like**
Update python3 specs in the cmsdist repository and in our requirements.txt text file.
CherryPy is a good candidate, but we should also check whether there are other python libraries that we can update to a newer stable release. Note that those libraries no longer need to support python2.

If possible, we should avoid any releases that came out in the past month or two, since those could still be buggy.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by vkuznet** [comment]
There is a whole set of libs which are outdated. Here is the latest versions of python libraries listed in `requirements.txt` file:
- SQLAlchemy 1.4.37
- CherryPy 18.6.1
- cx-Oracle 8.3.0
- httplib2 0.20.4
- jinja2 3.1.2
- memory-profiler 0.60.0
- mysqlclient 2.1.0
- nose2 0.11.0
- psutil 5.9.1
- pycurl 7.45.1
- pylint 2.14.1
- pymongo 4.1.1
- pyOpenSSL 22.0.0
- pyzmq 23.1.0
- retry 0.9.3
- Sphinx 5.0.1
- SQLAlchemy 1.4.37

I don't know if you want to update `requirements.txt` with all newer versions of packages and have single PR for cmsdist or by individual packages. And, it is unclear how to move forward with testing different versions. It is always tedious process and certainly some guidelines are required to decide how to move forward with versions. Moreover, the version changes all the time and it is probably would be a good idea to make one single effort to update package versions once per quarter or semi annually.

**Comment by amaltaro** [comment]
This issue is catching my attention lately as well, but I'd rather upgrade CouchDB first, run the new setup stably for a month or so, and then upgrade all these dependencies. Of course, it does not mean that we cannot have things ready to go before that actually happens.

> I don't know if you want to update requirements.txt with all newer versions of packages and have single PR for cmsdist or by individual packages.

I think we need to create 1 PR for WMCore with the requirements.txt update AND link to its description the 1 cmsdist PR upgrading all these packages. 

> And, it is unclear how to move forward with testing different versions. It is always tedious process and certainly some guidelines are required to decide how to move forward with versions. Moreover, the version changes all the time and it is probably would be a good idea to make one single effort to update package versions once per quarter or semi annually.

Yes, that's why we should avoid updating dependencies separately, otherwise there is TONS of validation to be done.
However, we need to be careful with libraries with major upgrade, since those usually impose breaking changes.

If you want to come up with some initial guidelines, my take on this is more or less the following:
* we should avoid upgrading packages to the very latest version (especially if it's been recently released, chances of having a bug in there not yet fixed is large)
* if there is a major version upgrade, we might want to upgrade that library completely separated from anything else
* a summary of the changes notes should be provided in the PR (with things that might be relevant to us, like deprecation/removal, incompatibilities, perhaps new features)
* I'd suggest to exercise it twice a year, but if we managed to get it done only once, it's much better than what we have had over the past many years!
* when providing these new versions, a full integration validation must be performed
* when involving SQLAlchemy library, validation with both MariaDB and Oracle databases should be done (WMAgent)

**Comment by amaltaro** [comment]
Note that specific tickets were created for some of the third-party python libraries (which might require baseline changes in the WMCore code):
https://github.com/dmwm/WMCore/issues/11205
https://github.com/dmwm/WMCore/issues/11206
https://github.com/dmwm/WMCore/issues/11207

@mapellidario @belforte I believe you are still on board with us in terms of relying on these cmsdist specs, right? If so, a well advanced friendly heads up that we will likely work on these updates during this quarter.

**Comment by belforte** [comment]
thanks Alan, yes we are with you on those !

**Comment by amaltaro** [comment]
Just a note: CherryPy 18.8.0 has been released a couple of days ago and it looks like a good candidate for python3-only projects. From the release notes:
https://github.com/cherrypy/cherrypy/blob/main/CHANGES.rst

I don't see any changes that would negatively affect us.

**Comment by amaltaro** [comment]
Deploying HG2209a-comp4, which brings the latest python third-party libraries provided in the cmsdist link above, fails to start the services with the following missing dependency:
```
+ sudo -H -u _workqueue bashs -lc '/data/srv/current/config/workqueue/manage status '\''I did read documentation'\'''
Traceback (most recent call last):
  File "/data/srv/HG2209a-comp4/sw/slc7_amd64_gcc630/cms/workqueue/2.1.2.pre1/bin/wmc-httpd", line 2, in <module>
    from WMCore.REST.Main import main
  File "/data/srv/HG2209a-comp4/sw/slc7_amd64_gcc630/cms/workqueue/2.1.2.pre1/lib/python3.8/site-packages/WMCore/REST/Main.py", line 33, in <module>
    import cherrypy
  File "/data/srv/HG2209a-comp4/sw/slc7_amd64_gcc630/external/py3-cherrypy/18.8.0/lib/python3.8/site-packages/cherrypy/__init__.py", line 66, in <module>
    from ._cperror import (
  File "/data/srv/HG2209a-comp4/sw/slc7_amd64_gcc630/external/py3-cherrypy/18.8.0/lib/python3.8/site-packages/cherrypy/_cperror.py", line 135, in <module>
    from cherrypy.lib import httputil as _httputil
  File "/data/srv/HG2209a-comp4/sw/slc7_amd64_gcc630/external/py3-cherrypy/18.8.0/lib/python3.8/site-packages/cherrypy/lib/httputil.py", line 20, in <module>
    import jaraco.collections
ModuleNotFoundError: No module named 'jaraco.collections'
```

@vkuznet Valentin, can you please look into it and provide another cmsdist PR fixing it (don't forget to link it to this GH issue)?

According to my test, the current CherryPy version requires the following packages/versions:
```
autocommand-2.2.1 cheroot-8.6.0 importlib-resources-5.9.0 inflect-6.0.0 jaraco.classes-3.2.2 jaraco.collections-3.5.2
jaraco.context-4.1.2 jaraco.functools-3.5.1 jaraco.text-3.9.1 more-itertools-8.14.0 portend-3.1.0 pydantic-1.9.2
pytz-2022.2.1 six-1.16.0 tempora-5.0.2 typing-extensions-4.3.0 zc.lockfile-2.0 zipp-3.8.1
```
even though it's not always that we need to create a spec for every single dependency (maybe modules that are never used in WMCore).

**Comment by vkuznet** [comment]
I added new PR here: https://github.com/cms-sw/cmsdist/pull/8043 , I included all jaraco packages and adjusted others. Tomorrow, I'll check the build status and we will proceed if I'll get time.
